src/librustc_trans/trans_item.rs

```diff
@@ -99,6 +102,35 @@ impl<'a, 'tcx> TransItem<'tcx> {
                ccx.codegen_unit().name);
     }
 
+    /// If necessary, creates a subtask for trans'ing a particular item and registers reads on
+    /// `TypeckItemBody` and `Hir`.
+    fn register_reads(&self, ccx: &CrateContext<'a, 'tcx>) {
+        let tcx = ccx.tcx();
+        let def_id = match *self {
+            TransItem::Static(node_id) => {
+                tcx.map.local_def_id(node_id)
+            }
+            TransItem::Fn(instance) => {
+                instance.def
+            }
+            TransItem::DropGlue(_) => {
+                // Nothing to track for drop glue
+                return;
+            }
+        };
+
+        tcx.dep_graph.with_task(DepNode::TransCrateItem(def_id), || {
+            tcx.dep_graph.read(DepNode::Hir(def_id));
+
+            // We are going to be accessing various tables
+            // generated by TypeckItemBody; we also assume
+            // that the body passes type check. These tables
+            // are not individually tracked, so just register
+            // a read here.
+            tcx.dep_graph.read(DepNode::TypeckItemBody(def_id));
```

Comment:
> But MIR is probably not tracked in the dep graph at all, is it now?

Surprisingly no. We seem to generate dependency nodes for various MIR-related things but I can't find a place where we register read-edges to those. If this is cleaned up, this explicit `register_reads` should not be necessary anymore. Not a problem for this PR though.